File: src/erlang/algorithms_2.erl
%% Definition for singly-linked list.
%%
%% -record(list_node, {val = 0 :: integer(),
%%                     next = null :: 'null' | #list_node{}}).

-spec add_two_numbers(L1 :: #list_node{} | null, L2 :: #list_node{} | null) -> #list_node{} | null.
add_two_numbers(L1, L2) ->
    F = fun(X, Y, C) ->
                Sum = X + Y + C,
                {Sum rem 10, Sum div 10}
        end,
    zip_with_list(L1, L2, 0, F).

-spec zip_with_list(L1 :: #list_node{} | null, L2 :: #list_node{} | null, Carry :: integer(),
                     F :: fun((integer(), integer(), integer()) -> {integer(), integer()})) -> #list_node{} | null.
zip_with_list(null, null, 0, _) ->
    null;
zip_with_list(L1, L2, Carry, F) ->
    A = case L1 of 
        null -> 0; 
        #list_node{val=V1} -> V1 
    end,
    B = case L2 of 
        null -> 0; 
        #list_node{val=V2} -> V2 
    end,

    {Digit, NextCarry} = F(A, B, Carry),

    Next1 = case L1 of 
        null -> null; 
        #list_node{next=N1} -> N1 
    end,
    Next2 = case L2 of 
        null -> null; 
        #list_node{next=N2} -> N2 
    end,
    
    #list_node{val=Digit, next=zip_with_list(Next1, Next2, NextCarry, F)}.

File: src/cpp/algorithms_2.cpp
#include <utility>
using namespace std;

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
    constexpr static auto next_of = [](auto node) { return node ? node->next : nullptr; };
    constexpr static auto val_of = [](auto node) { return node ? node->val : 0; };

    template<typename Func>
    ListNode* zip_list(ListNode* l1, ListNode* l2, Func F, int carry)  {
        if(l1 == nullptr && l2 == nullptr && carry == 0) return nullptr;

        const auto [digit, carry_new] = F(l1, l2, carry);

        return new ListNode(digit, zip_list(
            next_of(l1),
            next_of(l2),
            F,
            carry_new
        ));
    }

public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        auto F = [](ListNode* l1, ListNode* l2, int c) -> std::pair<int, int> {
            int sum = val_of(l1) + val_of(l2) + c;
            return {sum % 10, sum / 10};
        };

        return zip_list(l1, l2, F, 0);
    }
};